### ISSUE TITLE
Spawn emulator API31 hangs

### ISSUE THREAD

**Issue by ericschaal** [wix/Detox#3071]
### Describe the bug

Spawning an Android emulator hangs. 

`childProcessPromise` in `launchEmulatorProcess.js` never resolves.

### Steps To Reproduce

- [x] I have tested this issue on the latest Detox release and it still reproduces

Using the versions specified below, start detox making sure the emulator is off.

1. Detox 19.0.0
2. RN 0.66.2
3. Node v14.18.1
4. MacOS 12.0.1 (Intel)
5. Android Emulator 30.9.5, API 31

Detox hangs at ` DEBUG: [SPAWN_CMD] /home/eschaal/Android/Sdk/emulator/emulator -verbose -no-audio -no-boot-anim -port 19346 -gpu swiftshader_indirect @Pixel_4_API_31`

If the emulator is already up and running when starting detox, tests are executed as expected.

Looks like the spawn child process is not detached properly.
I was able to force childProcessPromise to resolve using this **bad workaround** in child-process-promise...

```
diff --git a/node_modules/child-process-promise/lib/index.js b/node_modules/child-process-promise/lib/index.js
index 2da7577..db6a818 100644
--- a/node_modules/child-process-promise/lib/index.js
+++ b/node_modules/child-process-promise/lib/index.js
@@ -105,6 +105,11 @@ function doSpawn(method, command, args, options) {
         }
     }
 
+    var detached = options && options.detached
+    if (detached) {
+        cp.on("spawn", () => setTimeout(() => resolve(0), 5000))
+    }
+
     result.childProcess = cp;
 
     if (captureStdout) {

```

After a quick search I noticed a similar issue happened with node 12.10 https://github.com/wix/Detox/issues/1546.

EDIT: Issue doesn't seem to affect my Android API 30 emulator. A `SPAWN_SUCCESS` is printed in the logs and tests start as expected.

### Expected behavior
Detox should check for `getprop dev.bootcomplete` and proceed with tests.

### Detox Trace-Logs

<details>
 <summary>Detox logs</summary>

```
detox[13697] DEBUG: [WSS_CREATE] Detox server listening on localhost:37013...
261detox[13697] DEBUG: [WSS_CONNECTION, #35082] registered a new connection.
262detox[13697] DEBUG: [EXEC_CMD, #0] "/home/eschaal/Android/Sdk/emulator/emulator" -list-avds --verbose
263detox[13697] DEBUG: [EXEC_CMD, #1] "/home/eschaal/Android/Sdk/emulator/emulator" -version 
264detox[13697] DEBUG: [EMU_BIN_VERSION_DETECT] Detected emulator binary version { major: 30, minor: 6, patch: 5, toString: [Function: toString] }
265detox[13697] DEBUG: [ALLOCATE_DEVICE] Trying to allocate a device based on "Pixel_4_API_31"
266detox[13697] DEBUG: [EXEC_CMD, #2] "/home/eschaal/Android/Sdk/platform-tools/adb"  devices
267detox[13697] DEBUG: [EXEC_SUCCESS, #2] List of devices attached
268detox[13697] DEBUG: [ALLOCATE_DEVICE] Settled on emulator-19346
269detox[13697] DEBUG: [SPAWN_CMD] /home/eschaal/Android/Sdk/emulator/emulator -verbose -no-audio -no-boot-anim -port 19346 -gpu swiftshader_indirect @Pixel_4_API_31
270detox[13697] ERROR: Exceeded timeout of 300000ms while handling jest-circus "setup" event
```

</details>

### Environment

 - Detox: 19
 - React Native: 0.66.2
 - Node: v14.18.1
 - Device: Pixel 4 Emulator  30.9.5, API 31
 - OS:  MacOS 12.0.1 (Intel)
 - Test-runner: jest-circus

**Comment by psysize** [comment]
Anyway we can open this up again? I've ran into the exact same issue with Detox 19.4.1.

@ericschaal's _bad_ workaround seems to work nicely with [patch-package](https://github.com/ds300/patch-package) though.

**Comment by d4vidi** [comment]
Interesting case. We're struggling ourselves with issues in that area - not with the bootstrap of emulators, but with ADB commands hanging indefinitely. Could you try to recreate the emulator from scratch and set up a working quickboot snapshot for it?

**Comment by matthewparavati** [comment]
Hello, not sure if this is the appropriate place/how much info I can add but this issue is affecting me on CircleCi only. Locally it is building and working fine. 

I've sent in a support ticket to CircleCi as well but saw this issue that sounds quite similar so thought I'd post here, too.

I've tried with both Android 30 and 31 on CircleCi but both result in the same issue. Before, our tests were working on Android 30. On the branch experiencing this issue, I _have_ upgraded a bunch of dependencies, including react-native to `0.69.4` and detox to `19.9.1`.

Any thoughts or additional info I could try providing?

I see the comment about
> Could you try to recreate the emulator from scratch and set up a working quickboot snapshot for it?

But I'm not exactly sure how I'd do that locally, let alone if I could try on CircleCi.

Environment
Detox: 19.9.1
React Native: 0.69.4
Node: v14.20.0
Device: Creating avd w/ `system-images;android-31;default;x86_64` or `system-images;android-30;default;x86_64`
OS: CircleCi's Linux machine executor
Test-runner: jest-circus

<img width="1446" alt="Screen Shot 2022-08-17 at 07 46 39" src="https://user-images.githubusercontent.com/12176557/185124075-808bfae7-548e-46ac-a968-8bc6ee525bfe.png">

**Comment by d4vidi** [comment]
@matthewparavati does it work if you prelaunch an emulator (e.g. using CircleCI's support for doing so, if any?)

**Comment by matthewparavati** [comment]
Hey @d4vidi thanks for the response. I haven't found anything that specifically mentions prelaunch but our process in our `config.yml` has been to create and start the emulator before we build the app and then run the tests. When starting the emulator, we also pass it the parameter `wait-for-emulator: true`

In circleci's android orb's `run-tests` command, I also don't see a prelaunch option. 
- We pass it a `test-command`, could we try and add something here to prelaunch the emulator? 
- There is also a `retry-interval` parameter, which is the "Duration in seconds to wait before the next try" 
  - This is currently set to the default value of 5 seconds. Not sure if increasing that will help. 

I did see the conversation in [#3342](https://github.com/wix/Detox/issues/3342) yesterday after I posted here and have now tried using API 28 and that seems to work on a branch where I've also made changes to utilize test-butler. I'm pretty sure I also tried API 29 (though I tried lots of things so it's all blurring together) and I don't think that worked either

**Comment by d4vidi** [comment]
@matthewparavati that sounds ok, I was suggesting to get an emulator ready before running Detox.
Anyways regarding the effect of API's on this - it's intriguing, please try to follow #3560.

**Comment by jonathanmos** [comment]
The problem is being investigated in #3636 - closing this one

**Comment by jonathanmos** [comment]
@ericschaal @psysize @matthewparavati I've released a fix for this issue in the latest detox (20.0.3) - please try it and see if this issue is now solved for you

**Comment by matthewparavati** [comment]
@jonathanmos thanks! Upgraded and was able to use API 31 on circleci for our E2E tests 🥳
